tools/coredump/gosym.go

```diff
@@ -0,0 +1,206 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/peterbourgon/ff/v3/ffcli"
+	"go.opentelemetry.io/ebpf-profiler/libpf"
+	"go.opentelemetry.io/ebpf-profiler/reporter"
+	"go.opentelemetry.io/ebpf-profiler/tools/coredump/modulestore"
+)
+
+type gosymCmd struct {
+	store    *modulestore.Store
+	casePath string
+}
+
+func newGosymCmd(store *modulestore.Store) *ffcli.Command {
+	args := &gosymCmd{store: store}
+
+	set := flag.NewFlagSet("gosym", flag.ExitOnError)
+	set.StringVar(&args.casePath, "case", "", "Path of the test case to debug")
+
+	return &ffcli.Command{
+		Name:       "gosym",
+		Exec:       args.exec,
+		ShortUsage: "gosym",
+		ShortHelp:  "Symbolize go test case",
+		FlagSet:    set,
+	}
+}
+
+func (cmd *gosymCmd) exec(context.Context, []string) (err error) {
+	// Validate arguments.
+	if cmd.casePath == "" {
+		return errors.New("please specify `-case`")
+	}
+
+	var test *CoredumpTestCase
+	test, err = readTestCase(cmd.casePath)
+	if err != nil {
+		return fmt.Errorf("failed to read test case: %w", err)
+	}
+
+	if got := len(test.Modules); got != 1 {
+		return fmt.Errorf("got=%d module but only 1 module is supported right now", got)
+	}
+
+	binary, err := extractModuleToTempFile(cmd.store, test.Modules[0])
+	if err != nil {
+		return fmt.Errorf("failed to extract binary: %w", err)
+	}
+	defer os.Remove(binary)
+
+	addrs := map[libpf.AddressOrLineno]struct{}{}
+	frames := map[libpf.AddressOrLineno][]*string{}
+	for _, thread := range test.Threads {
+		for i, frame := range thread.Frames {
+			_, addr, err := parseUnsymbolizedFrame(frame)
+			if err != nil {
+				continue
+			}
+			addrs[addr] = struct{}{}
+			frames[addr] = append(frames[addr], &thread.Frames[i])
+		}
+	}
+
+	locs, err := goSymbolize(binary, addrs)
+	if err != nil {
+		return fmt.Errorf("failed to symbolize: %w", err)
+	}
+
+	for addr, frame := range locs {
+		for _, frameS := range frames[addr] {
+			*frameS = formatSymbolizedFrame(frame, false) + " (" + *frameS + ")"
+		}
+	}
+
+	return writeTestCaseJSON(os.Stdout, test)
+}
+
+func extractModuleToTempFile(store *modulestore.Store, m ModuleInfo) (string, error) {
+	file, err := os.CreateTemp("", "")
+	if err != nil {
+		return "", err
+	}
+	return file.Name(), store.UnpackModuleToPath(m.Ref, file.Name())
+}
+
+func goSymbolize(binary string, addrs map[libpf.AddressOrLineno]struct{}) (map[libpf.AddressOrLineno]*reporter.FrameMetadataArgs, error) {
+	// Launch addr2line process.
+	addr2line := exec.Command("go", "tool", "addr2line", binary)
```

Comment:
Is there a specific reason to spawn a subprocess instead of using the [standard library's module](https://pkg.go.dev/debug/gosym)? I'm guessing the command may also resolve ELF symbols (instead of relying soley on gosym)?

Reply:
I went with addr2line because I vaguely (mis-)remembered that debug/gosym had some issues that addr2line doesn't have. But after checking the addr2line implementation itself, it seems like it actually uses the debug/gosym package as well. And [the issue](https://github.com/golang/go/issues/58474) I remembered wasn't about a defect in gosym, but rather some data that was not being exposed, but it doesn't matter here.

So I applied your suggestion in 6e99bdd48398aef579fffbec92bc0129b942699e and it's making the implementation a lot simpler as well. Thanks! 🙇